### DISCUSSION TITLE
Getting distorted images while saving it using Renderer::readPixels

### DISCUSSION THREAD

**Discussion by Pankaj003** [google/filament#5968] [Q&A]
Hi,

I am trying to save the rendered 3D object  using below function/algorithm, I am not getting expected output. Could anyone please check and suggest if I am doing anything wrong. This is happening on Mac OS with Opaque background, if I create the same with transparent background, i.e. with 4 channels then I gets the correct output. Looks like something wrong happens while reading pixels and saving image with 3(rgb) channels.

![SnapShotIssue](https://user-images.githubusercontent.com/8086656/185839805-96677fb4-279d-47e2-9a1c-be913f7332c4.png).

function:
void saveScreenshotMac(const QString& filepath, bool transparency, int DPI)
{
    using namespace filament;
    using namespace std::chrono_literals;

    int w = mFilamentWidget.width();
    int h = mFilamentWidget.height();

    const bool transparent = transparency;
    int n_channels = transparent ? 4 : 3;
    auto dataformat = transparent ? backend::PixelBufferDescriptor::PixelDataFormat::RGBA : backend::PixelBufferDescriptor::PixelDataFormat::RGB;
    auto qimageformat = transparent ? QImage::Format_RGBA8888 : QImage::Format_RGB888;
    const Viewport& vp = RenderingEngine::Get().MainView()->getViewport();

    mFilamentWidget.GetView().GetCamManipulator().updateCamera(RenderingEngine::Get().MainCamera());

    Renderer::ClearOptions new_clear_options = RenderingEngine::Get().ClearOptions();
    if (transparent) {
        new_clear_options.clearColor[3] = 0.0;
    }
    else {
        new_clear_options.clearColor[3] = 1.0;
    }
    RenderingEngine::Get().Renderer()->setClearOptions(new_clear_options);

    bool done = false;

    uint8_t* pixels = new uint8_t[vp.width * vp.height * n_channels];
    backend::PixelBufferDescriptor buffer(
        pixels,
        vp.width * vp.height * n_channels,
        dataformat,
        backend::PixelBufferDescriptor::PixelDataType::UBYTE,
        [](void* buffer, size_t size, void* capture_state) {
            CaptureState* state = static_cast<CaptureState*>(capture_state);
            const Viewport& v = state->view->getViewport();
            QImage img(reinterpret_cast<unsigned char*>(buffer), v.width, v.height, state->img_format);

            int dpm = state->DPI / 0.0254;
            img.setDotsPerMeterX(dpm);
            img.setDotsPerMeterY(dpm);

            img.save(state->filepath);

            // cleanup.
            delete[] static_cast<uint8_t*>(buffer);
            delete state;

            *(state->done) = true;
        }, new CaptureState{ RenderingEngine::Get().MainView(), filepath, qimageformat, w, h, DPI , &done });

    auto renderer = RenderingEngine::Get().Renderer();
      while (!renderer->beginFrame(RenderingEngine::Get().SwapChain())) {
        std::this_thread::sleep_for(8ms);
    }

    renderer->render(RenderingEngine::Get().MainView());
    renderer->readPixels(0, 0, vp.width, vp.height, std::move(buffer));
    renderer->endFrame();
    while (!done) {
        // Keep flushing the render buffer.  Filament requires the
        // command stream to execute while CPU bound operations are scheduled
        if (renderer->beginFrame(RenderingEngine::Get().SwapChain())) {
            renderer->render(RenderingEngine::Get().MainView());
            renderer->endFrame();
        }
        std::this_thread::sleep_for(8ms);
    }
    RenderingEngine::Get().Renderer()->setClearOptions(RenderingEngine::Get().ClearOptions());
}

**Comment by bejado** [comment]
Which backend are you using (OpenGL, Metal, or Vulkan)? Some backends do not natively support 3 channel textures so we do a conversion on the CPU, which could be buggy.

**Reply by romainguy** [reply]
It could also be a mismatch on the Qt side.

**Comment by Pankaj003** [comment]
@bejado, @romainguy 
Hi,
I am using OpenGL backend with filament version v1.23.0. Previously I was using filament v1.9.1 and there too face the same issue.
As far as Qt is concerned I don't think Qt is causing issue here. We are using same code on Windows with OpenGL backend and everything works fine.

Thanks

**Reply by bejado** [reply]
Are you able to try using the Metal backend on MacOS for a point of comparison?

**Comment by Pankaj003** [comment]
@bejado, @romainguy 
One information, if it helps, on Windows OS same code is not able to reproduce images with transparent(rbga) background but opaque(rbg) works fine and on Mac OS it is opposite, we have issue with opaque images, i.e. we gets good transparent image but not opaque one.
Both Windows and Mac have opposite behavior, is it something to do with OS? Is it handled in filament?
Please suggest
